repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher_test.go

```diff
@@ -110,6 +111,149 @@ func TestInitIpfsFetcher(t *testing.T) {
+		"/ip4/104.131.131.82/tcp/4001/p2p/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ"
+	],
+	"Migration": {
+		"DownloadSources": ["IPFS", "HTTP", "127.0.0.1", "https://127.0.1.1"],
+		"Keep": "cache"
+	},
+	"Peering": {
+		"Peers": [
+			{
+				"ID": "12D3KooWGC6TvWhfapngX6wvJHMYvKpDMXPb3ZnCZ6dMoaMtimQ5",
+				"Addrs": ["/ip4/127.0.0.1/tcp/4001", "/ip4/127.0.0.1/udp/4001/quic"]
+			}
+		]
+	}
+}
+`
+
+	noSuchDir := "no_such_dir-5953aa51-1145-4efd-afd1-a069075fcf76"
+	bootstrap, peers := readIpfsConfig(&noSuchDir)
+	if bootstrap != nil {
+		t.Error("expected nil bootstrap")
+	}
+	if peers != nil {
+		t.Error("expected nil peers")
+	}
+
+	tmpDir := makeConfig(testConfig)
+	defer os.RemoveAll(tmpDir)
+
+	bootstrap, peers = readIpfsConfig(nil)
+	if bootstrap != nil || peers != nil {
+		t.Fatal("expected nil ipfs config items")
+	}
+
+	bootstrap, peers = readIpfsConfig(&tmpDir)
+	if len(bootstrap) != 2 {
+		t.Fatal("wrong number of bootstrap addresses")
+	}
+	if bootstrap[0] != "/dnsaddr/bootstrap.libp2p.io/p2p/QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt" {
+		t.Fatal("wrong bootstrap address")
+	}
+
+	if len(peers) != 1 {
+		t.Fatal("wrong number of peers")
+	}
+
+	peer := peers[0]
+	if peer.ID.String() != "12D3KooWGC6TvWhfapngX6wvJHMYvKpDMXPb3ZnCZ6dMoaMtimQ5" {
+		t.Errorf("wrong ID for first peer")
+	}
+	if len(peer.Addrs) != 2 {
+		t.Error("wrong number of addrs for first peer")
+	}
+}
+
+func TestReadPartialIpfsConfig(t *testing.T) {
+	const (
+		configBadBootstrap = `
+{
+	"Bootstrap": "unreadable",
+	"Migration": {
+		"DownloadSources": ["IPFS", "HTTP", "127.0.0.1"],
+		"Keep": "cache"
+	},
+	"Peering": {
+		"Peers": [
+			{
+				"ID": "12D3KooWGC6TvWhfapngX6wvJHMYvKpDMXPb3ZnCZ6dMoaMtimQ5",
+				"Addrs": ["/ip4/127.0.0.1/tcp/4001", "/ip4/127.0.0.1/udp/4001/quic"]
+			}
+		]
+	}
+}
+`
+		configBadPeers = `
+{
+	"Bootstrap": [
+		"/dnsaddr/bootstrap.libp2p.io/p2p/QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt",
+		"/ip4/104.131.131.82/tcp/4001/p2p/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ"
+	],
+	"Migration": {
+		"DownloadSources": ["IPFS", "HTTP", "127.0.0.1"],
+		"Keep": "cache"
+	},
+	"Peering": "Unreadable-data"
+}
+`
+	)
+
+	tmpDir := makeConfig(configBadBootstrap)
+	defer os.RemoveAll(tmpDir)
+
+	bootstrap, peers := readIpfsConfig(&tmpDir)
+	if bootstrap != nil {
+		t.Fatal("expected nil bootstrap")
+	}
+	if len(peers) != 1 {
+		t.Fatal("wrong number of peers")
+	}
+	if len(peers[0].Addrs) != 2 {
+		t.Error("wrong number of addrs for first peer")
+	}
+	os.RemoveAll(tmpDir)
+
+	tmpDir = makeConfig(configBadPeers)
+	defer os.RemoveAll(tmpDir)
+
+	bootstrap, peers = readIpfsConfig(&tmpDir)
+	if peers != nil {
+		t.Fatal("expected nil peers")
+	}
+	if len(bootstrap) != 2 {
+		t.Fatal("wrong number of bootstrap addresses")
+	}
+	if bootstrap[0] != "/dnsaddr/bootstrap.libp2p.io/p2p/QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt" {
+		t.Fatal("wrong bootstrap address")
+	}
+}
+
+func makeConfig(configData string) string {
```

Comment:
Non-blocker: I can see the code was moved here and this is a general cleanup suggestion unrelated to this PR. Since this is only used in `_test`s, would it make sense to take `*testing.T` as an argument here and:
* fail the tests on error removing the `panic` statements
* use `t.TempDir` for constructing ephemeral directories which would eliminate the need for `defer os.RemoveAll`s across tests.

Reply:
I chose to `panic` because the failures there are not a failure of anything that is being tested, but a programming error.  Changed to use `t.Fatal` since that is more expected and helps readability.